### PR TITLE
OHRM5X-2097: Develop attachment download controller

### PR DESCRIPTION
**Pull request by DulsaraNethmin**

### CODE REVIEW

File: installer/Migration/V5_5_0/lang-string/claim.yaml [hunk 1]
```diff
@@ -1,67 +1,45 @@
 langStrings:
-  -
-    value: "Claim"
+  - value: "Claim"
```
Comment (ManulMax) [inline installer/Migration/V5_5_0/lang-string/claim.yaml:2]: revert these changes

File: installer/Migration/V5_5_0/permission/screens.yaml [hunk 2]
```diff
@@ -3,46 +3,46 @@
   url: viewClaimModule
   permissions:
     - { role: Admin, permission: { read: true } }
-    - { role: ESS, permission: { read: true, create: false, update: false, delete: false }}
+    - { role: ESS, permission: { read: true, create: false, update: false, delete: false },}
```
Comment (ManulMax) [inline installer/Migration/V5_5_0/permission/screens.yaml:6]: revert these changes

File: src/plugins/orangehrmClaimPlugin/Api/ClaimExpenseAPI.php [hunk 3]
```diff
@@ -1,4 +1,5 @@
 <?php
+
```
Comment (ManulMax) [inline src/plugins/orangehrmClaimPlugin/Api/ClaimExpenseAPI.php:2]: revert these changes

File: src/plugins/orangehrmClaimPlugin/Api/ClaimExpenseAPI.php [hunk 4]
```diff
@@ -256,6 +257,11 @@ public function getValidationRuleForCreate(): ParamRuleCollection
                 self::PARAMETER_EXPENSE_TYPE_ID,
                 new Rule(Rules::POSITIVE)
             ),
+            //            new ParamRule(
+            //                self::PARAMETER_AMOUNT,
+            //                new Rule(Rules::FLOAT_TYPE), //TODO:: handle decimal points
+            //                new Rule(Rules::MIN, [0])
+            //            ),
```
Comment (ManulMax) [inline src/plugins/orangehrmClaimPlugin/Api/ClaimExpenseAPI.php:264]: remove this

File: src/plugins/orangehrmClaimPlugin/Api/EmployeeClaimRequestAPI.php [hunk 5]
```diff
@@ -1,4 +1,5 @@
 <?php
+
```
Comment (ManulMax) [inline src/plugins/orangehrmClaimPlugin/Api/EmployeeClaimRequestAPI.php:2]: revert these changes

File: src/plugins/orangehrmClaimPlugin/Dao/ClaimDao.php [hunk 6]
```diff
@@ -521,4 +523,25 @@ public function getClaimRequestCount(ClaimRequestSearchFilterParams $myClaimRequ
     {
         return $this->getClaimRequestPaginator($myClaimRequestSearchFilterParams)->count();
     }
+
+    /**
+     * @param int $requestId
+     * @param int $attachId
+     * @return ClaimAttachment|null
+     * @throws DaoException
+     */
```
Comment (ManulMax) [inline src/plugins/orangehrmClaimPlugin/Dao/ClaimDao.php:532]:
```suggestion
    /**
     * @param int $requestId
     * @param int $attachId
     * @return ClaimAttachment|null
     */
```

File: src/plugins/orangehrmClaimPlugin/Service/ClaimService.php [hunk 7]
```diff
@@ -48,4 +54,26 @@ public function getReferenceId(): string
         $date = $this->getDateTimeHelper()->getNow()->format('Ymd');
         return $date . str_pad("$nextId", 7, 0, STR_PAD_LEFT);
     }
+
+    /**
+     * @param int $requestId
+     * @param int $attachId
+     * @return ClaimAttachment|null
+     * @throws DaoException
```
Comment (ManulMax) [inline src/plugins/orangehrmClaimPlugin/Service/ClaimService.php:62]: .

**Review by ManulMax — CHANGES_REQUESTED** [review]

File: src/plugins/orangehrmClaimPlugin/Controller/ClaimAttachmentController.php [hunk 8]
```diff
@@ -0,0 +1,72 @@
...
+     * @var ClaimService|null
+     */
+    protected ?ClaimService $claimService = null;
+
+    /**
+     * @return ClaimService
+     */
+    public function getClaimService(): ClaimService
+    {
+        if (!$this->claimService instanceof ClaimService) {
```
Comment (RajithaKumara) [inline src/plugins/orangehrmClaimPlugin/Controller/ClaimAttachmentController.php:40]: check this, use service container

File: src/plugins/orangehrmClaimPlugin/Controller/ClaimAttachmentController.php [hunk 9]
```diff
@@ -0,0 +1,72 @@
...
+     * @param Request $request
+     * @return Response
+     */
+    public function handle(Request $request): Response
+    {
+        $requestId = $request->attributes->get('requestId');
+        $attachId = $request->attributes->get('attachId');
+        $response = $this->getResponse();
+
+        if ($requestId && $attachId) {
```
Comment (RajithaKumara) [inline src/plugins/orangehrmClaimPlugin/Controller/ClaimAttachmentController.php:56]:
```suggestion
        if ($request->attributes->has('requestId') && $request->attributes->has('attachId')) {
```

File: src/plugins/orangehrmClaimPlugin/Dao/ClaimDao.php [hunk 10]
```diff
@@ -521,4 +523,24 @@ public function getClaimRequestCount(ClaimRequestSearchFilterParams $myClaimRequ
...
+    public function getClaimAttachmentFile(int $requestId, int $attachId): ?ClaimAttachment
+    {
+        try {
+            $criteria = ['requestId' => $requestId, 'attachId' => $attachId];
+
+            $claimAttachment = $this->getRepository(ClaimAttachment::class)->findOneBy($criteria);
+            if ($claimAttachment instanceof ClaimAttachment) {
+                return $claimAttachment;
+            }
+            return null;
```
Comment (RajithaKumara) [inline src/plugins/orangehrmClaimPlugin/Dao/ClaimDao.php:541]:
```suggestion
            return $this->getRepository(ClaimAttachment::class)->findOneBy($criteria);
```

File: src/plugins/orangehrmClaimPlugin/Dao/ClaimDao.php [hunk 11]
```diff
@@ -521,4 +523,24 @@ public function getClaimRequestCount(ClaimRequestSearchFilterParams $myClaimRequ
...
+        } catch (Exception $e) {
+            throw new DaoException($e->getMessage(), $e->getCode(), $e);
+        }
```
Comment (RajithaKumara) [inline src/plugins/orangehrmClaimPlugin/Dao/ClaimDao.php:544]:
```suggestion
```
No need try catch block

File: src/plugins/orangehrmClaimPlugin/Controller/ClaimAttachmentController.php (hunk 9 above)
Comment (RajithaKumara) [inline src/plugins/orangehrmClaimPlugin/Controller/ClaimAttachmentController.php:56]:
```suggestion
        if ($request->attributes->has('requestId') && $request->attributes->has('attachId')) {
```

File: src/plugins/orangehrmClaimPlugin/Service/ClaimService.php [hunk 12]
```diff
@@ -19,12 +19,17 @@
...
 use OrangeHRM\Claim\Dao\ClaimDao;
 use OrangeHRM\Core\Traits\Service\DateTimeHelperTrait;
+use OrangeHRM\Core\Traits\UserRoleManagerTrait;
+use OrangeHRM\Entity\ClaimAttachment;
 
 class ClaimService
 {
     use DateTimeHelperTrait;
+    use UserRoleManagerTrait;
+    use ClaimRequestAPIHelperTrait;
```
Comment (RajithaKumara) [inline src/plugins/orangehrmClaimPlugin/Service/ClaimService.php:32]: remove this usage

File: src/plugins/orangehrmClaimPlugin/Service/ClaimService.php [hunk 13]
```diff
@@ -48,4 +53,25 @@ public function getReferenceId(): string
...
+    }
+
+    /**
+     * @param int $requestId
+     * @param int $attachId
+     * @return ClaimAttachment|null
+     */
+    public function getAccessibleClaimAttachment(int $requestId, int $attachId): ?ClaimAttachment
+    {
+        $this->getClaimRequest($requestId);
```
Comment (RajithaKumara) [inline src/plugins/orangehrmClaimPlugin/Service/ClaimService.php:74]: *

File: src/plugins/orangehrmClaimPlugin/entity/Decorator/ClaimAttachmentDecorator.php [hunk 14]
```diff
@@ -58,4 +63,16 @@ public function setUserByUserId(int $userId): void
...
         $this->getClaimAttachment()->setUser($user);
     }
+
+
+    /**
+     * @return string
+     */
+    public function getAttachment(): string
+    {
+        if (is_null($this->attachmentString)) {
```
Comment (RajithaKumara) [inline src/plugins/orangehrmClaimPlugin/entity/Decorator/ClaimAttachmentDecorator.php:73]:
```suggestion
        if ($this->attachmentString === null) {
```

**Review by RajithaKumara — CHANGES_REQUESTED** [review]

File: src/client/src/orangehrmClaimPlugin/pages/myClaims/MyClaims.vue [hunk 15]
```diff
@@ -153,7 +153,8 @@ export default {
     const serializedFilters = computed(() => {
       return {
         referenceId:
-          typeof filters.value.referenceId === 'object'
+          typeof filters.value.referenceId === 'object' &&
+          filters.value.referenceId !== null
             ? filters.value.referenceId.id
             : typeof filters.value.referenceId === 'string'
             ? filters.value.referenceId
```
Comment (RajithaKumara) [inline src/client/src/orangehrmClaimPlugin/pages/myClaims/MyClaims.vue:160]: @ManulMax check this
cc @Super-Chama

File: src/plugins/orangehrmClaimPlugin/Controller/ClaimAttachmentController.php [hunk 16]
```diff
@@ -19,42 +19,49 @@
...
+
+            if (!$this->getUserRoleManagerHelper()->isEmployeeAccessible($claimRequest->getEmployee()->getEmpNumber())) {
+                throw $this->getBadRequestException();
+            }
+
+            try {
+                $attachment = $this->getClaimService()->getClaimDao()
+                    ->getClaimAttachmentFile($requestId, $attachId);
+            } catch (Exception $e) {
+                throw $this->getBadRequestException();
```
Comment (RajithaKumara) [inline src/plugins/orangehrmClaimPlugin/Controller/ClaimAttachmentController.php:62]:
```suggestion
                return $this->handleBadRequest();
```

File: src/plugins/orangehrmClaimPlugin/Controller/ClaimAttachmentController.php [hunk 17]
```diff
@@ -19,42 +19,49 @@
...
+        if ($request->attributes->has('requestId') && $request->attributes->has('attachId')) {
+            $requestId = $request->attributes->get('requestId');
+            $attachId = $request->attributes->get('attachId');
+
+            $claimRequest = $this->getClaimService()->getClaimDao()
+                ->getClaimRequestById($requestId);
+            $this->throwRecordNotFoundExceptionIfNotExist($claimRequest, ClaimRequest::class);
+
+            if (!$this->getUserRoleManagerHelper()->isEmployeeAccessible($claimRequest->getEmployee()->getEmpNumber())) {
+                throw $this->getBadRequestException();
```
Comment (RajithaKumara) [inline src/plugins/orangehrmClaimPlugin/Controller/ClaimAttachmentController.php:55]:
```suggestion
                return $this->handleBadRequest();
```

File: src/plugins/orangehrmClaimPlugin/Controller/ClaimAttachmentController.php [hunk 18]
```diff
@@ -19,42 +19,49 @@
...
 
-        if ($requestId && $attachId) {
-            $attachment = $this->getClaimService()->getAccessibleClaimAttachment($requestId, $attachId);
+        if ($request->attributes->has('requestId') && $request->attributes->has('attachId')) {
+            $requestId = $request->attributes->get('requestId');
+            $attachId = $request->attributes->get('attachId');
+
+            $claimRequest = $this->getClaimService()->getClaimDao()
+                ->getClaimRequestById($requestId);
+            $this->throwRecordNotFoundExceptionIfNotExist($claimRequest, ClaimRequest::class);
```
Comment (RajithaKumara) [inline src/plugins/orangehrmClaimPlugin/Controller/ClaimAttachmentController.php:52]:
```suggestion
            return $this->handleBadRequest();
```

File: src/plugins/orangehrmClaimPlugin/Controller/ClaimAttachmentController.php [hunk 19]
```diff
@@ -19,42 +19,49 @@
...
-     */
-    public function getClaimService(): ClaimService
-    {
-        if (!$this->claimService instanceof ClaimService) {
-            $this->claimService = new ClaimService();
-        }
-        return $this->claimService;
-    }
+    use ClaimServiceTrait;
+    use EndpointExceptionTrait;
```
Comment (RajithaKumara) [inline src/plugins/orangehrmClaimPlugin/Controller/ClaimAttachmentController.php:35]:
```suggestion
```

File: src/plugins/orangehrmClaimPlugin/Controller/ClaimAttachmentController.php [hunk 20]
```diff
@@ -19,42 +19,49 @@
 
 namespace OrangeHRM\Claim\Controller;
 
-use OrangeHRM\Claim\Service\ClaimService;
+use Exception;
+use OrangeHRM\Claim\Traits\Service\ClaimServiceTrait;
+use OrangeHRM\Core\Api\V2\Exception\EndpointExceptionTrait;
```
Comment (RajithaKumara) [inline src/plugins/orangehrmClaimPlugin/Controller/ClaimAttachmentController.php:24]:
```suggestion
```
Don't use API helpers in controllers.
@ManulMax check this

**Review by RajithaKumara — CHANGES_REQUESTED** [review]

File: src/client/src/orangehrmClaimPlugin/pages/myClaims/MyClaims.vue (hunk 15 above)
Comment (ManulMax) [reply src/client/src/orangehrmClaimPlugin/pages/myClaims/MyClaims.vue:160]: Checked I think need to change ReferenceIdAutocomplete. Need a support from @Super-Chama.

Comment (ManulMax) [reply src/client/src/orangehrmClaimPlugin/pages/myClaims/MyClaims.vue:160]: ** In **ClaimRequestSearchFilterParams**, only added referenceId as allowed sort field. So as like other places we cannot search by id when he select the autocomplete result. That's why he had to get referenceId here in every time.

Comment (DulsaraNethmin) [reply src/client/src/orangehrmClaimPlugin/pages/myClaims/MyClaims.vue]: is this ok ?
typeof filters.value.referenceId === 'object' &&
          filters.value.referenceId?.id !== undefined
